### DISCUSSION TITLE
Tree: Ability to provide extra data per node, and include that data elsewhere

### DISCUSSION THREAD

**Discussion by ronnyek** [mantinedev/mantine#6670] [Feature requests]
So I've got the need to render a select tree... where I can check/uncheck items. Later I'd like to get a list of checked items and do things with it.

Currently as long as the data implements at LEAST `TreeNodeData` interface, the tree can render just fine... and `getCheckedNodes` will return the nodes and their select status. The problem seems to be that the `getCheckedNodes` always seems to produce `TreeNodeData`, which omits any additional data.

I'm not sure if a great implementation would be that TreeNodeData could just have a data property with string/object dictionary or something, and that data would just be included on TreeNodeData, which seemingly would require very little else.

Another option could be to have TreeNodeData have a generic parameter, and the `getCheckedNodes` methods and other methods that return nodes, could be utilized in a generic way such that you always get `TreeNodeData<T>`. 

If there is already a better way to do this, I'd love to know what that should look like. My guesses as to how best to implement this would be based on treeviews I've found in other ui toolkits.

**Comment by thomaslecoeur** [comment]
agree
